partiql-lang/src/main/kotlin/org/partiql/lang/planner/transforms/PartiQLSchemaInferencer.kt

```diff
@@ -0,0 +1,174 @@
+                        SourceLocationMeta(0, 0),
+                        PlanningProblemDetails.CompileError("Unhandled exception occurred.")
+                    ),
+                    cause = t
+                )
+            }
+        }
+    }
+
+    /**
+     * Context object required for performing schema inference.
+     */
+    public class Context(
+        public val session: PlannerSession,
+        plugins: List<Plugin>,
+        public val problemHandler: ProblemHandler = ProblemThrower()
+    ) {
+        internal val metadata = Metadata(plugins, session.catalogConfig)
+    }
+
+    public class InferenceException(
+        message: String = "",
+        errorCode: ErrorCode,
+        errorContext: PropertyValueMap,
+        cause: Throwable? = null
+    ) : SqlException(message, errorCode, errorContext, cause) {
+
+        constructor(err: Problem, cause: Throwable? = null) :
+            this(
+                message = "",
+                errorCode = ErrorCode.INTERNAL_ERROR,
+                errorContext = propertyValueMapOf(
+                    Property.LINE_NUMBER to err.sourceLocation.lineNum,
+                    Property.COLUMN_NUMBER to err.sourceLocation.charOffset,
+                    Property.MESSAGE to err.details.message
+                ),
+                cause = cause
+            )
+    }
+
+    //
+    //
+    // INTERNAL
+    //
+    //
+
+    internal class ProblemThrower : ProblemHandler {
+        override fun handleProblem(problem: Problem) {
+            if (problem.details.severity == ProblemSeverity.ERROR) {
+                throw InferenceException(problem)
+            }
+        }
+    }
+
+    private const val DEFAULT_TABLE_NAME = "UNSPECIFIED"
+
+    private fun inferInternal(query: String, ctx: Context): ValueDescriptor {
+        val parser = PartiQLParserBuilder.standard().build()
+        val ast = parser.parseAstStatement(query)
+        val normalizedAst = ast.normalize()
+        return inferUsingLogicalPlan(normalizedAst, ctx)
+    }
+
+    /**
+     * Infers using the logical plan.
+     */
+    private fun inferUsingLogicalPlan(normalizedAst: PartiqlAst.Statement, ctx: Context): ValueDescriptor {
+        val query = normalizedAst as PartiqlAst.Statement.Query
+        val plan = when (val expr = query.expr) {
+            is PartiqlAst.Expr.Select -> RelConverter.convert(expr)
+            else -> RexConverter.convert(expr)
+        }
+        val typedPlan = PlanTyper.type(
+            plan,
+            PlanTyper.Context(
+                input = null,
+                session = ctx.session,
+                metadata = ctx.metadata,
+                scopingOrder = PlanTyper.ScopingOrder.LEXICAL_THEN_GLOBALS,
+                customFunctionSignatures = emptyList(),
+                problemHandler = ctx.problemHandler
+            )
+        )
+        return convertSchema(typedPlan)
+    }
+
+    private fun convertSchema(rex: Rex): ValueDescriptor = when (rex) {
```

Comment:
Do you see the `valueDescriptor` getting replaced by a schema Type? If so, may be mark this in a comment?